crypto/fipsmodule/rand/fork_detect.c

```diff
@@ -42,37 +42,102 @@ DEFINE_STATIC_MUTEX(g_fork_detect_lock)
 DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
 DEFINE_BSS_GET(uint64_t, g_fork_generation)
 DEFINE_BSS_GET(int, g_ignore_madv_wipeonfork)
+DEFINE_BSS_GET(int, g_ignore_pthread_atfork)
 
+static int init_fork_detect_madv_wipeonfork(void *addr, long page_size) {
+
+  // Some versions of qemu (up to at least 5.0.0-rc4, see linux-user/syscall.c)
+  // ignore |madvise| calls and just return zero (i.e. success). But we need to
+  // know whether MADV_WIPEONFORK actually took effect. Therefore try an invalid
+  // call to check that the implementation of |madvise| is actually rejecting
+  // unknown |advice| values.
+  if (madvise(addr, (size_t)page_size, -1) == 0 ||
+      madvise(addr, (size_t)page_size, MADV_WIPEONFORK) != 0) {
+    munmap(addr, (size_t)page_size);
+    return 0;
+  }
+
+  return 1;
+}
+
+static void pthread_atfork_on_fork(void) {
+
+  struct CRYPTO_STATIC_MUTEX *const lock = g_fork_detect_lock_bss_get();
+
+  // Aquire locks to be on the safe side. We want to avoid the checks in
```

Comment:
I suggest putting at the beginning of this comment:
This handler reinforces the same mechanism of detecting a fork as `MADV_WIPEONFORK`: it sets to 0 the same location that should have been part of the wipe and which is checked in `CRYPTO_get_fork_detection()`.

Reply:
Pushed - slightly modified text.